### PR TITLE
Client/generic resource interface

### LINKED ISSUE #115: [client] Refactor client package to accept generic Resource interface
After discussion in #67, it was agreed to use `Common interface for resource` option for actual implementation.

Possible new types:
```go

type (
	// ResourceAction represents possible action on resource
	ResourceAction string

	// Resource interface declares a contract for abstract RouterOS resource
	Resource interface {
		// ActionToCommand trnaslates CRUD action to RouterOS command path
		ActionToCommand(ResourceAction) string

		// AfterAddHook is called right after the resource successfully added
		// This hook is mainly used to set resource's ID field based on reply from RouterOS
		AfterAddHook(r *routeros.Reply)

		// IDField reveals name of ID field to use in requests to MikroTik router
		// It is used in operations like Find
		IDField() string
		// ID returns value of the ID field
		ID() string
		// SetID updates a value of the ID field
		SetID(string)

		// DeleteField retrieves a name of a field which is used for resource deletion
		DeleteField() string

		// SetDeleteID sets value of the DeleteField field to be used during deletion of the resource
		SetDeleteID(string)
	}
)

```

### PR DESCRIPTION
**Pull request by maksym-nazarenko**
This PR introduces new set of interfaces to segregate `Client` implementation from resources. Because of that, resource could be more declarative and actual logic of manipulating lands inside client itself.

The interfaces list is:
```go
const (
	Add    Action = "add"
	Update Action = "update"
	Find   Action = "find"
	List   Action = "list"
	Delete Action = "delete"
)

type (
	// Action represents possible action on resource
	Action string

	// Resource interface defines a contract for abstract RouterOS resource
	Resource interface {
		// ActionToCommand trnaslates CRUD action to RouterOS command path
		ActionToCommand(Action) string

		// IDField reveals name of ID field to use in requests to MikroTik router
		// It is used in operations like Find
		IDField() string

		// ID returns value of the ID field
		ID() string

		// SetID updates a value of the ID field
		SetID(string)
	}

	// Adder defines contract for resources which require custom behaviour during resource creation
	Adder interface {
		// AfterAddHook is called right after the resource successfully added
		// This hook is mainly used to set resource's ID field based on reply from RouterOS
		AfterAddHook(r *routeros.Reply)
	}

	// Finder defines contract for resources which provide custom behaviour during resource retrieval
	Finder interface {
		// FindField retrieves a name of a field to use as key for resource retrieval
		FindField() string

		// FindFieldValue retrieves a value to use for resource retrieval
		FindFieldValue() string
	}

	// Deleter defines contract for resources which require custom behaviour during resource deletion
	Deleter interface {
		// DeleteField retrieves a name of a field which is used for resource deletion
		DeleteField() string

		// DeleteFieldValue retrieves a value for DeleteField field
		DeleteFieldValue() string
	}
)
```

It is intentionally split into smaller interfaces to let developers omit some of them and rely on default behavior.

This PR also includes first implementation of the interfaces for `DnsRecord`.

Closes #115

### CODE REVIEW

**Comment by ddelnano** [comment]
I'll be giving this a 👀  soon

**Comment by ddelnano** [comment]
Thanks for working on this new pattern! Only two minor comments and one thought for as we onboard more resources to this pattern.

**Comment by maksym-nazarenko** [comment]
@ddelnano have you submitted your review or I got you wrong?
I don't see  
>Only two minor comments and one thought for as we onboard more resources to this pattern.

**Comment by ddelnano** [comment]
@maksym-nazarenko sorry! You are right that I forgot press the submit button.

File: client/client_crud.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	log.Printf("[DEBUG] find response: %v", r)
+
+	targetStruct := client.newTargetStruct(d)
+	targetStructInterface := targetStruct.Interface()
+	err = Unmarshal(*r, targetStructInterface)
+	if err != nil {
+		return nil, err
+	}
+	if targetStructInterface.(Resource).ID() == "" {
+		return nil, NewNotFound(fmt.Sprintf("resource with field `%s=%s` not found", field, value))
```
[the remark this answers is not in the document]
Comment (maksym-nazarenko) [reply client/client_crud.go]: done

File: client/client_crud.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		deleteFieldValue = deleter.DeleteFieldValue()
+	}
+	cmd := []string{d.ActionToCommand(Delete), "=" + deleteField + "=" + deleteFieldValue}
+	log.Printf("[INFO] Running the mikrotik command: `%s`", cmd)
+	_, err = c.RunArgs(cmd)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (maksym-nazarenko) [reply client/client_crud.go]: done

File: client/dns_crud.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	return d.Name
+}
+
+func (d *DnsRecord) DeleteField() string {
+	return "numbers"
+}
+
+func (d *DnsRecord) DeleteFieldValue() string {
+	return d.Id
+}
```
[the remark this answers is not in the document]
Comment (maksym-nazarenko) [reply client/dns_crud.go:45]: I'm not sure if it will be the case, but will keep it in mind

**Comment by maksym-nazarenko** [comment]
@ddelnano could we also keep this PR or create an issue to discuss further steps (or use discord channel of this repo)?
The questions would be:
1. do we want to migrate all resources to this new interface?
2. should we keep both ways of implementing client resources, remove old or mark old as deprecated?
3. describe best practices for new way ...

**Comment by ddelnano** [comment]
Copying a link to my reply on [Discord](https://discord.com/channels/704471459279667330/705514478019543122/1039069032273420338).

> I don't have a strong opinion one way or the other. The new interface will make it easy to avoid common mistakes that happen when implementing a new resource with the existing pattern, so I think it should gradually shift to the new interface over time

The PR I just merged created a conflict for this one. Once that is resolved, this is ready to be merged.

File: go.work [hunk 4]
```diff
@@ -0,0 +1,6 @@
+go 1.19
+
+use (
+	.
+	./client
+)
```
Comment (ddelnano) [inline go.work:6]: Nice, didn't know there was specific support for muti module workspaces 👍

**Pull request merged**